core/src/main/java/org/pentaho/di/core/fileinput/CharsetToolkit.java

```diff
@@ -0,0 +1,441 @@
+          // a four-bytes sequence was encoutered
+          // there must be three continuation bytes of the form 10xxxxxx,
+          // otherwise the following characteris is not a valid UTF-8 construct
+          if ( !( isContinuationChar( b1 ) && isContinuationChar( b2 ) && isContinuationChar( b3 ) ) ) {
+            validU8Char = false;
+          } else {
+            i += 3;
+          }
+        } else if ( isFiveBytesSequence( b0 ) ) {
+          // a five-bytes sequence was encoutered
+          // there must be four continuation bytes of the form 10xxxxxx,
+          // otherwise the following characteris is not a valid UTF-8 construct
+          if ( !( isContinuationChar( b1 ) && isContinuationChar( b2 )
+            && isContinuationChar( b3 ) && isContinuationChar( b4 ) ) ) {
+            validU8Char = false;
+          } else {
+            i += 4;
+          }
+        } else if ( isSixBytesSequence( b0 ) ) {
+          // a six-bytes sequence was encoutered
+          // there must be five continuation bytes of the form 10xxxxxx,
+          // otherwise the following characteris is not a valid UTF-8 construct
+          if ( !( isContinuationChar( b1 )
+            && isContinuationChar( b2 ) && isContinuationChar( b3 )
+            && isContinuationChar( b4 ) && isContinuationChar( b5 ) ) ) {
+            validU8Char = false;
+          } else {
+            i += 5;
+          }
+        } else {
+          validU8Char = false;
+        }
+      }
+      if ( !validU8Char ) {
+        break;
+      }
+      i++;
+    }
+    // if no byte with an high order bit set, the encoding is US-ASCII
+    // (it might have been UTF-7, but this encoding is usually internally used only by mail systems)
+    if ( !highOrderBit ) {
+      // returns the default charset rather than US-ASCII if the enforce8Bit flag is set.
+      if ( this.enforce8Bit ) {
+        return this.defaultCharset;
+      } else {
+        return Charset.forName( "US-ASCII" );
+      }
+    }
+    // if no invalid UTF-8 were encountered, we can assume the encoding is UTF-8,
+    // otherwise the file would not be human readable
+    if ( validU8Char ) {
+      return Charset.forName( "UTF-8" );
+    }
+    // finally, if it's not UTF-8 nor US-ASCII, let's assume the encoding is the default encoding
+    return this.defaultCharset;
+  }
+
+  public static Charset guessEncoding( File f, int bufferLength ) throws FileNotFoundException, IOException {
+    FileInputStream fis = new FileInputStream( f );
+    byte[] buffer = new byte[bufferLength];
+    fis.read( buffer );
+    fis.close();
+    CharsetToolkit toolkit = new CharsetToolkit( buffer );
+    toolkit.setDefaultCharset( getDefaultSystemCharset() );
+    return toolkit.guessEncoding();
+  }
+
+  public static Charset guessEncoding( File f, int bufferLength, Charset defaultCharset ) throws FileNotFoundException, IOException {
+    FileInputStream fis = new FileInputStream( f );
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**
This line of code contains a resource that might not be closed properly. This can cause a resource leak that slows down or crashes your system.

**Fix**
Consider closing the following resource: *fis*. The resource is referenced at line 317. The resource is closed at line 318. There are other execution paths that do not contain closure statements, for example, when FileInputStream.read throws an exception. Ensure *fis* is closed in a try-finally block or declared in a try-with-resources block.

**More info**
[View resource management guidelines at oracle.com](https://www.oracle.com/technetwork/articles/java/trywithresources-401775.html) (external link).